<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-NAN" data-toc-modified-id="Удаление-NAN-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Удаление NAN</a></span></li><li><span><a href="#Удаление-колонок" data-toc-modified-id="Удаление-колонок-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Удаление колонок</a></span></li><li><span><a href="#One-hot-encoding" data-toc-modified-id="One-hot-encoding-1.2.3"><span class="toc-item-num">1.2.3&nbsp;&nbsp;</span>One hot encoding</a></span></li><li><span><a href="#Разделение-выборки" data-toc-modified-id="Разделение-выборки-1.2.4"><span class="toc-item-num">1.2.4&nbsp;&nbsp;</span>Разделение выборки</a></span></li><li><span><a href="#Масштабирование-выборки" data-toc-modified-id="Масштабирование-выборки-1.2.5"><span class="toc-item-num">1.2.5&nbsp;&nbsp;</span>Масштабирование выборки</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Определение-баланса-классов" data-toc-modified-id="Определение-баланса-классов-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Определение баланса классов</a></span></li><li><span><a href="#Обучение-моделей-на-несбалансированной-выборке" data-toc-modified-id="Обучение-моделей-на-несбалансированной-выборке-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обучение моделей на несбалансированной выборке</a></span><ul class="toc-item"><li><span><a href="#DecisionTreeClassifier" data-toc-modified-id="DecisionTreeClassifier-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>DecisionTreeClassifier</a></span></li><li><span><a href="#RandomForestClassifier" data-toc-modified-id="RandomForestClassifier-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>RandomForestClassifier</a></span></li><li><span><a href="#LogisticRegression" data-toc-modified-id="LogisticRegression-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>LogisticRegression</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Сlass-weight.-Обучение-моделей-на-сбалансированной-выборке." data-toc-modified-id="Сlass-weight.-Обучение-моделей-на-сбалансированной-выборке.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Сlass weight. Обучение моделей на сбалансированной выборке.</a></span><ul class="toc-item"><li><span><a href="#DecisionTreeClassifier" data-toc-modified-id="DecisionTreeClassifier-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>DecisionTreeClassifier</a></span></li><li><span><a href="#RandomForestClassifier" data-toc-modified-id="RandomForestClassifier-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>RandomForestClassifier</a></span></li><li><span><a href="#LogisticRegression" data-toc-modified-id="LogisticRegression-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>LogisticRegression</a></span></li></ul></li><li><span><a href="#Upsampling.-Обучение-моделей-на-сбалансированной-выборке." data-toc-modified-id="Upsampling.-Обучение-моделей-на-сбалансированной-выборке.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Upsampling. Обучение моделей на сбалансированной выборке.</a></span><ul class="toc-item"><li><span><a href="#DecisionTreeClassifier" data-toc-modified-id="DecisionTreeClassifier-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>DecisionTreeClassifier</a></span></li><li><span><a href="#RandomForestClassifier" data-toc-modified-id="RandomForestClassifier-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>RandomForestClassifier</a></span></li><li><span><a href="#LogisticRegression" data-toc-modified-id="LogisticRegression-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>LogisticRegression</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Тестирование-лучшей-модели" data-toc-modified-id="Тестирование-лучшей-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование лучшей модели</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Прогнозирование оттока клиента банка

Нам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком «Бета-Банка».

**Цель проекта** - спрогнозировать, уйдёт клиент из банка в ближайшее время или нет.


**План работы:**
1. Подготовка данных.
2. Исследование данных.
3. Определение дисбаланса данных.
4. Обучение моделей без/с учетом дисбаланса.
5. Тестирование лучшей модели.
6. Определение метрик F1 и AUC-ROC.

## Подготовка данных

### Обзор данных

Импортируем необходимые библиотеки и данные.

In [1]:
# Импорт стандартных библиотек
import pandas as pd
from sklearn.utils import shuffle
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score, roc_auc_score
from sklearn.model_selection import train_test_split

In [2]:
try:
    data = pd.read_csv('D:\\Users\\BlackEdition\\Desktop\\Data Science\\'
                       'Обучение в Яндекс\\8. Обучение с учителем\\Проект\\Churn.csv')
except:
    data = pd.read_csv('/datasets/Churn.csv')

Проведем их обзор.

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [4]:
data.head(10)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   
5     8.0  113755.78              2          1               0   
6     7.0       0.00              2          1               1   
7     4.0  115046.74              4          1               0   
8     4.0  142051.07              2          0               1   
9     2.0  134603.88              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

**Описание данных:**
* `RowNumber` — индекс строки в данных;
* `CustomerId` — уникальный идентификатор клиента;
* `Surname` — фамилия;
* `CreditScore` — кредитный рейтинг;
* `Geography` — страна проживания;
* `Gende`r — пол;
* `Age` — возраст;
* `Tenure` — сколько лет человек является клиентом банка;
* `Balance` — баланс на счёте;
* `NumOfProducts` — количество продуктов банка, используемых клиентом;
* `HasCrCard` — наличие кредитной карты;
* `IsActiveMember` — активность клиента;
* `EstimatedSalary` — предполагаемая зарплата;
* `Exited` — факт ухода клиента.

Из данных видно, что нам необходимо будет:
1. Удалить NAN в колонке *`Tenure`*.
2. Удалить колонки *`RowNumber`, `CustomerId`, `Surname`*, являющиеся лишними для моделей обучения.
3. Преобразовать категориальные данные в количественные.
4. Произвести разлеление данных на обучающую, валидационную и тестовую выборки.
5. Произвести масштабирование разделенных выборок.

### Предобработка данных

#### Удаление NAN

Найдем NAN объекты.

In [5]:
data.isna().sum()

RowNumber            0
CustomerId           0
Surname              0
CreditScore          0
Geography            0
Gender               0
Age                  0
Tenure             909
Balance              0
NumOfProducts        0
HasCrCard            0
IsActiveMember       0
EstimatedSalary      0
Exited               0
dtype: int64

Удалим NAN объекты, т.к. их не много, а их замена на количественные объекты может привести к ухудшению обучаемости моделей.

In [6]:
data = data.dropna()
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        9091 non-null   int64  
 1   CustomerId       9091 non-null   int64  
 2   Surname          9091 non-null   object 
 3   CreditScore      9091 non-null   int64  
 4   Geography        9091 non-null   object 
 5   Gender           9091 non-null   object 
 6   Age              9091 non-null   int64  
 7   Tenure           9091 non-null   float64
 8   Balance          9091 non-null   float64
 9   NumOfProducts    9091 non-null   int64  
 10  HasCrCard        9091 non-null   int64  
 11  IsActiveMember   9091 non-null   int64  
 12  EstimatedSalary  9091 non-null   float64
 13  Exited           9091 non-null   int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.0+ MB


#### Удаление колонок

Удалим колонки *RowNumber, CustomerId, Surname*, т.к. они бесполезны для обучения.

In [7]:
# Удаление ненужных столбцов
data = data.drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)
data.columns

Index(['CreditScore', 'Geography', 'Gender', 'Age', 'Tenure', 'Balance',
       'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary',
       'Exited'],
      dtype='object')

#### One hot encoding

Приведем категориальные данные к количественным с помощью One-Hot Encoding.

In [8]:
data_ohe = pd.get_dummies(data, drop_first=True)
data_ohe.dtypes

CreditScore            int64
Age                    int64
Tenure               float64
Balance              float64
NumOfProducts          int64
HasCrCard              int64
IsActiveMember         int64
EstimatedSalary      float64
Exited                 int64
Geography_Germany      uint8
Geography_Spain        uint8
Gender_Male            uint8
dtype: object

#### Разделение выборки

Разделим данные на фичи и целевые признаки.

In [9]:
# Разделение датасета на фичи и целевые признаки
features = data_ohe.drop('Exited', axis=1)
target = data_ohe['Exited']

display(features.shape)
display(target.shape)

(9091, 11)

(9091,)

Разобьем фичи и целевые признаки на обучающую, валидационну и тестовую выборки.

In [10]:
# Разделение выборки на обучающую и валидационную
(features_train, features_valid_test, target_train,
 target_valid_test) = (train_test_split(features,
                                        target,
                                        test_size=.4,
                                        random_state=12345))
# Разделение валидационной выборки на валидационную и тестовую
(features_valid, features_test, target_valid,
 target_test) = (train_test_split(features_valid_test,
                                  target_valid_test,
                                  test_size=.5,
                                  random_state=12345))

display(features_train.shape)
display(target_train.shape)

display(features_valid.shape)
display(target_valid.shape)

display(features_test.shape)
display(target_test.shape)

(5454, 11)

(5454,)

(1818, 11)

(1818,)

(1819, 11)

(1819,)

#### Масштабирование выборки

Обучим скейлер на обучающей выборке и масштабируем остальные выборки.

In [11]:
# Исключение ошибки взаимодействия фрагментов фреймов данных sklearn и pandas
pd.options.mode.chained_assignment = None

numeric = ['CreditScore', 'Age', 'Tenure', 'Balance', 'EstimatedSalary']

# Обучение скейлера на обучающей выборке для масштабирования остальных выборок
scaler = StandardScaler()
scaler.fit(features_train[numeric])

features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])

In [12]:
display(features_train)
display(features_valid)
display(features_test)

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
9344     0.809075 -1.039327 -1.025995  0.554904              1          1   
3796    -1.152518 -1.227561  0.696524  0.480609              1          0   
7462    -0.398853  0.090079  1.385532 -1.237830              1          1   
1508    -0.749875 -0.286389  0.352020 -1.237830              2          1   
4478    -1.028628 -0.756975 -0.336987 -1.237830              2          0   
...           ...       ...       ...       ...            ...        ...   
4938     1.283988 -0.474624  1.730036  0.813031              1          0   
4529     0.158652  0.466548  1.041028  1.071609              1          1   
3866    -1.049276 -0.945210  0.352020  0.449165              1          1   
2414    -0.956359  1.219485  1.385532 -1.237830              1          1   
5056     1.222043 -1.509912 -1.025995  1.015938              1          0   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
9344               0         0.019508                  0                0   
3796               0         0.056167                  0                0   
7462               1         0.848738                  0                0   
1508               1        -0.894953                  0                0   
4478               1        -1.284516                  0                0   
...              ...              ...                ...              ...   
4938               0        -1.092767                  1                0   
4529               1         0.260842                  0                0   
3866               1        -1.086965                  0                0   
2414               0         1.288527                  0                0   
5056               1        -1.244002                  1                0   

      Gender_Male  
9344            0  
3796            1  
7462            1  
1508            1  
4478            1  
...           ...  
4938            1  
4529            0  
3866            0  
2414            0  
5056            1  

[5454 rows x 11 columns]

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
7445    -1.369326  0.560665 -0.336987 -1.237830              1          1   
8620     1.232367  0.090079  1.041028 -1.237830              2          0   
1714     0.840048  0.560665  0.352020  1.231363              1          0   
5441     1.056856 -0.945210 -1.370498  0.951231              1          1   
9001     0.406433 -0.662858  0.352020  0.767800              1          1   
...           ...       ...       ...       ...            ...        ...   
7505     0.107031 -1.133444 -0.336987 -1.237830              2          1   
2232     0.003790 -0.286389  1.041028 -1.237830              2          1   
3720    -1.813265  0.090079  1.041028  1.226442              2          1   
5196     0.726482 -1.604030  0.696524  0.610110              1          1   
2460     0.117356  0.184196 -1.370498  0.860215              1          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
7445               0        -0.086537                  0                0   
8620               1        -0.537457                  0                0   
1714               0         1.070393                  1                0   
5441               0        -0.576279                  0                0   
9001               1         0.662068                  0                1   
...              ...              ...                ...              ...   
7505               0        -0.016504                  0                1   
2232               0         1.056449                  0                1   
3720               0        -0.468713                  1                0   
5196               1         1.328951                  0                0   
2460               0        -1.278650                  1                0   

      Gender_Male  
7445            0  
8620            0  
1714            1  
5441            1  
9001            1  
...           ...  
7505            1  
2232            1  
3720            1  
5196            0  
2460            1  

[1818 rows x 11 columns]

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
5170     1.707279 -0.756975 -0.336987 -1.237830              2          1   
4180    -0.429826 -0.286389  1.730036  0.485538              2          0   
7349    -0.171721  0.278313  0.352020 -0.269213              2          0   
7469     0.385784 -0.380507  1.041028  0.464813              1          1   
3467    -1.142194  0.278313 -1.370498  0.353837              2          0   
...           ...       ...       ...       ...            ...        ...   
8579     0.014114 -0.474624 -0.336987  0.474274              1          0   
7536     0.499350  0.278313 -1.370498 -1.237830              1          1   
2911     0.003790  1.219485  1.041028 -0.444590              1          1   
2789     0.581944  0.466548 -1.715002  1.742887              1          1   
7755     0.272218 -0.474624  1.041028 -1.237830              2          1   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
5170               1         0.718362                  0                0   
4180               0         1.687305                  1                0   
7349               1         0.824128                  1                0   
7469               0        -1.118018                  0                0   
3467               1        -0.107642                  1                0   
...              ...              ...                ...              ...   
8579               1         0.734995                  0                0   
7536               0        -1.704943                  0                0   
2911               0        -1.006086                  0                0   
2789               0        -0.541247                  0                1   
7755               1         1.455393                  0                0   

      Gender_Male  
5170            0  
4180            1  
7349            1  
7469            0  
3467            0  
...           ...  
8579            1  
7536            0  
2911            0  
2789            1  
7755            1  

[1819 rows x 11 columns]

### Вывод

В данном разделе мы:
- Удалили NAN в колонке `Tenure`.
- Удалили колонки `RowNumber`, `CustomerId`, `Surname`, являющиеся лишними для моделей обучения.
- Преобразовали категориальные данные в количественные.
- Произвели разлеление данных на обучающую, валидационную и тестовую выборки.
- Произвели масштабирование разделенных выборок.

Все поставленные задачи выполнены.

## Исследование задачи

### Определение баланса классов

Определим баланс классов.

In [13]:
# Определение дисбаланса классов
target_train_zero = target_train[target_train == 0]
target_train_one = target_train[target_train == 1]

display('Unbalanced training data: ',
        target_train_one.count() / target_train_zero.count() * 100, '%')

'Unbalanced training data: '

26.01663585951941

'%'

Видно, что класс 1 в выборке составляет 26% от класса 0 (в 3.8 раз меньше).

### Обучение моделей на несбалансированной выборке

Обучим модели на несбалансированной выборке и посмотрим на метрики качества.

#### DecisionTreeClassifier

Обучение Решающего дерева.

In [14]:
%%time

best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

# Обучение модели с перебором глубины
for depth in range(1, 10):
    model = DecisionTreeClassifier(max_depth=depth,
                                   random_state=12345)
    model.fit(features_train, target_train)

    predicted_valid = model.predict(features_valid)

    f1_valid = f1_score(target_valid, predicted_valid)

    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

    if f1_valid > best_score:
        best_score = f1_valid
        best_depth = depth
        best_model_score_auc_roc = auc_roc

    if auc_roc > best_auc_roc:
        best_auc_roc = auc_roc
        best_auc_roc_depth = depth

display(f'Best depth: {best_depth},best score: {best_score}, AUC-ROCK: {best_model_score_auc_roc}')
display(f'Best AUC-ROC depth: {best_auc_roc_depth}, best AUC-ROC: {best_auc_roc}')

'Best depth: 7,best score: 0.5764331210191083, AUC-ROCK: 0.8346049843812412'

'Best AUC-ROC depth: 6, best AUC-ROC: 0.8495399046859958'

Wall time: 318 ms


#### RandomForestClassifier

Обучение Рандомного леса.

In [15]:
%%time

best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

# Обучение модели с перебором глубины
for depth in range(1, 21):
    model = RandomForestClassifier(n_estimators=10,
                                   max_depth=depth,
                                   random_state=12345)
    model.fit(features_train, target_train)

    predicted_valid = model.predict(features_valid)

    f1_valid = f1_score(target_valid, predicted_valid)

    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

    if f1_valid > best_score:
        best_score = f1_valid
        best_depth = depth
        best_mscore_auc_roc = auc_roc

    if auc_roc > best_auc_roc:
        best_auc_roc = auc_roc
        best_auc_roc_depth = depth

display(f'Best depth: {best_depth}, best score: {best_score}, AUC-ROCK: {best_mscore_auc_roc}')
display(f'Best AUC-ROC depth: {best_auc_roc_depth}, best AUC-ROC: {best_auc_roc}')

'Best depth: 19, best score: 0.5780730897009967, AUC-ROCK: 0.8440931271939431'

'Best AUC-ROC depth: 9, best AUC-ROC: 0.8694172371455866'

Wall time: 1.94 s


In [16]:
%%time

best_est = 0  # Количество деревьев у модели с лучшей F1 метрикой
best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_est = 0  # Количество деревьев у лучшей AUC-ROC метрики
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

for est in range(1, 102, 10):
    display(f'Estimators: {est}')
    for depth in range(1, 25):
        model = RandomForestClassifier(n_estimators=est,
                                       max_depth=depth,
                                       random_state=12345)
        model.fit(features_train, target_train)

        predicted_valid = model.predict(features_valid)

        f1_valid = f1_score(target_valid, predicted_valid)

        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]

        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

        if f1_valid > best_score:
            best_score = f1_valid
            best_depth = depth
            best_est = est
            best_model_score_auc_roc = auc_roc

        if auc_roc > best_auc_roc:
            best_auc_roc = auc_roc
            best_auc_roc_depth = depth
            best_auc_roc_est = est

display(f'Best estimators: {best_est}, best depth: {best_depth},'
        f'best score: {best_score}, AUC-ROC: {best_model_score_auc_roc}')
display(
    f'Best AUC-ROC est: {best_auc_roc_est}, best AUC-ROCK depth {best_auc_roc_depth},'
    f'best AUC-ROC: {best_auc_roc}')

'Estimators: 1'

'Estimators: 11'

'Estimators: 21'

'Estimators: 31'

'Estimators: 41'

'Estimators: 51'

'Estimators: 61'

'Estimators: 71'

'Estimators: 81'

'Estimators: 91'

'Estimators: 101'

'Best estimators: 81, best depth: 23,best score: 0.602291325695581, AUC-ROC: 0.8616925322637352'

'Best AUC-ROC est: 81, best AUC-ROCK depth 9,best AUC-ROC: 0.8739381176864724'

Wall time: 1min 54s


#### LogisticRegression

Обучение логистической регрессии.

In [17]:
%%time

model = LogisticRegression(solver='liblinear', random_state=12345)
model.fit(features_train, target_train)

predicted_valid = model.predict(features_valid)

f1_valid = f1_score(target_valid, predicted_valid)

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

display(f'F1 valid {f1_valid}')
display(f'AUC-ROC: {auc_roc}')

'F1 valid 0.30400000000000005'

'AUC-ROC: 0.7735454858381945'

Wall time: 52.5 ms


### Вывод

В данном разделе мы обучили три модели на несбалансированной выборке и посмотрели на метрики их качества. Видно, что при обучении без учета дисбаланса классов, налучший показатель **F1 = 0.6023** у **RandomForest** с количеством деревьев **81** и глубиной **23**. **AUC-ROC** данной модели составляет **0.8617**.

## Борьба с дисбалансом

### Сlass weight. Обучение моделей на сбалансированной выборке.

Сбалансируем классы путем добавления в модели гиперпараментра class_weght.

#### DecisionTreeClassifier

Обучение Решающего дерева с гиперпараметром class_weight='balanced'.

In [18]:
%%time

# Переменные для фиксации лучших значений
best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

for depth in range(1, 10):
    model = DecisionTreeClassifier(max_depth=depth,
                                   random_state=12345,
                                   class_weight='balanced')
    model.fit(features_train, target_train)

    predicted_valid = model.predict(features_valid)

    f1_valid = f1_score(target_valid, predicted_valid)

    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

    if f1_valid > best_score:
        best_score = f1_valid
        best_depth = depth
        best_model_score_auc_roc = auc_roc

    if auc_roc > best_auc_roc:
        best_auc_roc = auc_roc
        best_auc_roc_depth = depth

display(f'Best depth: {best_depth}, best score: {best_score}, AUC-ROCK: {best_model_score_auc_roc}')
display(f'Best AUC-ROC depth: {best_auc_roc_depth}, best AUC-ROC: {best_auc_roc}')

'Best depth: 5, best score: 0.5735449735449736, AUC-ROCK: 0.8396523192522141'

'Best AUC-ROC depth: 5, best AUC-ROC: 0.8396523192522141'

Wall time: 232 ms


#### RandomForestClassifier

Обучение Рандомного леса с гиперпараметром class_weight='balanced'.

In [19]:
%%time

best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

for depth in range(1, 21):
    model = (RandomForestClassifier(n_estimators=10,
                                    max_depth=depth,
                                    random_state=12345,
                                    class_weight='balanced'))
    model.fit(features_train, target_train)

    predicted_valid = model.predict(features_valid)

    f1_valid = f1_score(target_valid, predicted_valid)

    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

    if f1_valid > best_score:
        best_score = f1_valid
        best_depth = depth
        best_model_score_auc_roc = auc_roc

    if auc_roc > best_auc_roc:
        best_auc_roc = auc_roc
        best_auc_roc_depth = depth

display(f'Best depth: {best_depth}, best score: {best_score}, AUC-ROCK: {best_model_score_auc_roc}')
display(f'Best AUC-ROC depth: {best_auc_roc_depth}, best AUC-ROC: {best_auc_roc}')

'Best depth: 9, best score: 0.6497584541062802, AUC-ROCK: 0.8690573706367336'

'Best AUC-ROC depth: 9, best AUC-ROC: 0.8690573706367336'

Wall time: 1.73 s


In [20]:
%%time

best_est = 0  # Количество деревьев у модели с лучшей F1 метрикой
best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_est = 0  # Количество деревьев у лучшей AUC-ROC метрики
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

for est in range(1, 112, 10):
    display(f'Estimators: {est}')
    for depth in range(1, 25):
        model = (RandomForestClassifier(n_estimators=est,
                                        max_depth=depth,
                                        random_state=12345,
                                        class_weight='balanced'))
        model.fit(features_train, target_train)

        predicted_valid = model.predict(features_valid)

        f1_valid = f1_score(target_valid, predicted_valid)

        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]

        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

        if f1_valid > best_score:
            best_score = f1_valid
            best_depth = depth
            best_est = est
            best_model_score_auc_roc = auc_roc

        if auc_roc > best_auc_roc:
            best_auc_roc = auc_roc
            best_auc_roc_depth = depth
            best_auc_roc_est = est

display(f'Best estimators: {best_est}, best depth: {best_depth},'
        f'best score: {best_score}, AUC-ROC: {best_model_score_auc_roc}')

display((
    f'Best AUC-ROC est: {best_auc_roc_est}, best AUC-ROCK depth {best_auc_roc_depth},'
    f'best AUC-ROC: {best_auc_roc}'))

'Estimators: 1'

'Estimators: 11'

'Estimators: 21'

'Estimators: 31'

'Estimators: 41'

'Estimators: 51'

'Estimators: 61'

'Estimators: 71'

'Estimators: 81'

'Estimators: 91'

'Estimators: 101'

'Estimators: 111'

'Best estimators: 101, best depth: 8,best score: 0.6460807600950118, AUC-ROC: 0.8695939490885529'

'Best AUC-ROC est: 31, best AUC-ROCK depth 9,best AUC-ROC: 0.8715874806951406'

Wall time: 2min 12s


#### LogisticRegression

Обучение Логистической регрессии с гиперпараметром class_weight='balanced'.

In [21]:
%%time

model = LogisticRegression(solver='liblinear',
                           random_state=12345,
                           class_weight='balanced')
model.fit(features_train, target_train)

predicted_valid = model.predict(features_valid)

f1_valid = f1_score(target_valid, predicted_valid)

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

display(f'F1 valid {f1_valid}')
display(f'AUC-ROC: {auc_roc}')

'F1 valid 0.509731232622799'

'AUC-ROC: 0.7778123429610663'

Wall time: 30.4 ms


**Лучшей моделью** по итогам обучения на сбалансированной выборке **является Рандомный лес** с **количеством деревьев 101** и их **глубиной 8**. **F1** метрика при этом составляет **0.6461**, а **AUC-ROC** - **0.8617**.

### Upsampling. Обучение моделей на сбалансированной выборке.

Сбалансируем обучающую выборку с помощью метода upsampling.

In [22]:
# Функция балансировки даннах методом unsampling
def upsample(features, target):
    # Repeat определяется путем частного между большим и меньшим классами
    repeat = round(target[target == 0].count() / target[target == 1].count())
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    # Upsampling
    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)

    # Перемешивание выборок
    features_upsampled, target_upsampled = (shuffle(features_upsampled,
                                                    target_upsampled,
                                                    random_state=12345))
    return features_upsampled, target_upsampled


features_upsampled, target_upsampled = upsample(features_train, target_train)

display(features_train.shape, features_upsampled.shape)
display(target_train.shape, target_upsampled.shape)

(5454, 11)

(8832, 11)

(5454,)

(8832,)

#### DecisionTreeClassifier

Обучение Решающего дерева на upsampled выборке.

In [23]:
%%time

best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

for depth in range(1, 10):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(features_upsampled, target_upsampled)

    predicted_valid = model.predict(features_valid)

    f1_valid = f1_score(target_valid, predicted_valid)

    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

    if f1_valid > best_score:
        best_score = f1_valid
        best_depth = depth
        best_model_score_auc_roc = auc_roc

    if auc_roc > best_auc_roc:
        best_auc_roc = auc_roc
        best_auc_roc_depth = depth

display(f'Best depth: {best_depth}, best score: {best_score}, AUC-ROCK: {best_model_score_auc_roc}')
display(f'Best AUC-ROC depth: {best_auc_roc_depth}, best AUC-ROC: {best_auc_roc}')

'Best depth: 5, best score: 0.5735449735449736, AUC-ROCK: 0.8396523192522141'

'Best AUC-ROC depth: 5, best AUC-ROC: 0.8396523192522141'

Wall time: 268 ms


#### RandomForestClassifier

Обучение Рандомного леса на upsampled выборке.

In [24]:
%%time

best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

for depth in range(1, 21):
    model = RandomForestClassifier(n_estimators=10,
                                   max_depth=depth,
                                   random_state=12345)
    model.fit(features_upsampled, target_upsampled)

    predicted_valid = model.predict(features_valid)

    f1_valid = f1_score(target_valid, predicted_valid)

    probabilities_valid = model.predict_proba(features_valid)
    probabilities_one_valid = probabilities_valid[:, 1]

    auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

    if f1_valid > best_score:
        best_score = f1_valid
        best_depth = depth
        best_model_score_auc_roc = auc_roc

    if auc_roc > best_auc_roc:
        best_auc_roc = auc_roc
        best_auc_roc_depth = depth

display(f'Best depth: {best_depth}, best score: {best_score}, AUC-ROCK: {best_model_score_auc_roc}')
display(f'Best AUC-ROC depth: {best_auc_roc_depth}, best AUC-ROC: {best_auc_roc}')

'Best depth: 9, best score: 0.6121076233183858, AUC-ROCK: 0.8550612325289871'

'Best AUC-ROC depth: 6, best AUC-ROC: 0.8576355205731356'

Wall time: 2.26 s


In [25]:
%%time

best_est = 0  # Количество деревьев у модели с лучшей F1 метрикой
best_depth = 0  # Лучшая глубина
best_score = 0  # Лучшая F1 метрика
best_model_score_auc_roc = 0  # Лучшая AUC-ROC у модели с лучшей F1 метрикой

best_auc_roc = 0  # Определение лучшего AUC-ROC значения
best_auc_roc_est = 0  # Количество деревьев у лучшей AUC-ROC метрики
best_auc_roc_depth = 0  # Глубина лучшего AUC-ROC значения

for est in range(1, 102, 10):
    display(f'Estimators: {est}')
    for depth in range(1, 25):
        model = RandomForestClassifier(n_estimators=est,
                                       max_depth=depth,
                                       random_state=12345)
        model.fit(features_upsampled, target_upsampled)

        predicted_valid = model.predict(features_valid)

        f1_valid = f1_score(target_valid, predicted_valid)

        probabilities_valid = model.predict_proba(features_valid)
        probabilities_one_valid = probabilities_valid[:, 1]

        auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

        if f1_valid > best_score:
            best_score = f1_valid
            best_depth = depth
            best_est = est
            best_model_score_auc_roc = auc_roc

        if auc_roc > best_auc_roc:
            best_auc_roc = auc_roc
            best_auc_roc_depth = depth
            best_auc_roc_est = est

display((f'Best estimators: {best_est}, best depth: {best_depth},'
         f'best score: {best_score}, AUC-ROC: {best_model_score_auc_roc}'))
display((
    f'Best AUC-ROC est: {best_auc_roc_est}, best AUC-ROCK depth {best_auc_roc_depth},'
    f'best AUC-ROC: {best_auc_roc}'))

'Estimators: 1'

'Estimators: 11'

'Estimators: 21'

'Estimators: 31'

'Estimators: 41'

'Estimators: 51'

'Estimators: 61'

'Estimators: 71'

'Estimators: 81'

'Estimators: 91'

'Estimators: 101'

'Best estimators: 81, best depth: 14,best score: 0.6378378378378378, AUC-ROC: 0.8614707219603246'

'Best AUC-ROC est: 101, best AUC-ROCK depth 8,best AUC-ROC: 0.8701369701632929'

Wall time: 2min 26s


#### LogisticRegression

Обучение Логистической регрессии на upsampled выборке.

In [26]:
%%time

model = LogisticRegression(solver='liblinear', random_state=12345)
model.fit(features_upsampled, target_upsampled)

predicted_valid = model.predict(features_valid)

f1_valid = f1_score(target_valid, predicted_valid)

probabilities_valid = model.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

display(f'F1 valid {f1_valid}')
display(f'AUC-ROC: {auc_roc}')

'F1 valid 0.5085972850678734'

'AUC-ROC: 0.7779246286748261'

Wall time: 38.5 ms


**Лучшей моделью** по итогам обучения на сбалансированной выборке **является Рандомный лес** с **количеством деревьев 81** и их **глубиной 14**. **F1** метрика при этом составляет **0.6378**, а **AUC-ROC** - **0.8617**.

### Вывод

По итогам проведенных мероприятий по балансировке классов обучающей выборки лучший показатель метрики F1 составляет 0.6461 у Рандомного леса с гиперпараметрами:
* `estimator: 101`;
* `depth: 8`;
* `class_weight: 'balanced'`.

AUC-ROC лучшей модели составляет 0.8617.

## Тестирование лучшей модели

Проведем тестирование лучшей модели (Рандомный лес с гиперпараметром `class_weight='balanced'`) на тестовой выборке.

In [27]:
%%time

model = (RandomForestClassifier(n_estimators=101,
                                max_depth=8,
                                random_state=12345,
                                class_weight='balanced'))
model.fit(features_train, target_train)

predicted_test = model.predict(features_test)

f1_test = f1_score(target_test, predicted_test)

probabilities_test = model.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]

auc_roc = roc_auc_score(target_test, probabilities_one_test)

display(f'Best score: {f1_test}, AUC-ROC: {auc_roc}')

'Best score: 0.5945273631840796, AUC-ROC: 0.8617631989566595'

Wall time: 778 ms


### Вывод

По итогам тестирования лучшей модели по итогам балансировки классов на тестовой выборке получены следующие метрики качества:
* **F1: 0.5945**
* **AUC-ROCK: 0.8618**

## Общий вывод

Нами были изучены и обработаны исторические данные о поведении клиентов и расторжении договоров с банком «Бета-Банка». Данные подготовлены для обучения математических моделей: проведено их численное категорирование и масштабирование. Обучены модели без учета дисбаланса классов. 

В данных мы наблюдаем дисбаланс классов, поэтому для повышения качества наших моделей нами была проведена их балансировка. По итогам балансировки проведено повторное обучение моделей с целью получить более точные предсказания.

По итогам проведенных мероприятий по балансировке классов обучающей выборки лучший показатель метрики F1 составляет 0.6461 у Рандомного леса с гиперпараметрами:

* `estimators: 101`;
* `depth: 8`;
* `class_weight: 'balanced'`.

AUC-ROC лучшей модели составляет 0.8617.

По итогам тестирования лучшей модели по итогам балансировки классов на тестовой выборке получены следующие метрики качества:

* F1: 0.5945
* AUC-ROCK: 0.8618